src/limiter/maximum_principle_limiter.cpp

```diff
@@ -0,0 +1,279 @@
+        const unsigned int n_dofs_curr_cell = current_fe_ref.n_dofs_per_cell();
+        // Obtain the mapping from local dof indices to global dof indices
+        current_dofs_indices.resize(n_dofs_curr_cell);
+        soln_cell->get_dof_indices(current_dofs_indices);
+
+        // Extract the local solution dofs in the cell from the global solution dofs
+        if (this->global_max.size() < nstate && this->global_min.size() < nstate) {
+            for (unsigned int istate = 0; istate < nstate; ++istate) {
+                this->global_max.push_back(-1e9);
+                this->global_min.push_back(1e9);
+            }
+        }
+
+        // Allocate solution dofs and set global max and min
+        std::array<std::vector<real>, nstate> soln_dofs;
+        const unsigned int n_shape_fns = n_dofs_curr_cell / nstate;
+        for (unsigned int idof = 0; idof < n_dofs_curr_cell; ++idof) {
+            const unsigned int istate = fe_collection[poly_degree].system_to_component_index(idof).first;
+            const unsigned int ishape = fe_collection[poly_degree].system_to_component_index(idof).second;
+            if (ishape == 0) {
+                soln_dofs[istate].resize(n_shape_fns);
+            }
+            soln_dofs[istate][ishape] = solution[current_dofs_indices[idof]]; //
+            if (soln_dofs[istate][ishape] > this->global_max[istate])
+                this->global_max[istate] = soln_dofs[istate][ishape];
+            if (soln_dofs[istate][ishape] < this->global_min[istate])
+                this->global_min[istate] = soln_dofs[istate][ishape];
+        }
+    }
+
+    // Print the obtained values for verification
+    for (unsigned int istate = 0; istate < nstate; ++istate) {
+        std::cout << std::fixed;
+        std::cout << std::setprecision(14);
+        std::cout << "global_max:   " << this->global_max[istate] << "   global_min:   " << this->global_min[istate] << std::endl;
+    }
+}
+
+template <int dim, int nstate, typename real>
+std::array<real, nstate> MaximumPrincipleLimiter<dim, nstate, real>::get_soln_cell_avg(
+    std::array<std::vector<real>, nstate> soln_at_q,
+    const unsigned int n_quad_pts,
+    const std::vector<real>& quad_weights)
+{
+    std::array<real, nstate> soln_cell_avg;
+    for (unsigned int istate = 0; istate < nstate; ++istate) {
+        soln_cell_avg[istate] = 0;
+    }
+
+    // Apply integral for solution cell average (dealii quadrature operates from [0,1])
+    for (unsigned int istate = 0; istate < nstate; ++istate) {
+        for (unsigned int iquad = 0; iquad < n_quad_pts; ++iquad) {
+            soln_cell_avg[istate] += quad_weights[iquad]
+                * soln_at_q[istate][iquad];
+        }
+    }
+
+    return soln_cell_avg;
+}
+
+template <int dim, int nstate, typename real>
+void MaximumPrincipleLimiter<dim, nstate, real>::write_limited_solution(
+    dealii::LinearAlgebra::distributed::Vector<double>      solution,
```

Comment:
Good catch on function arguments @PranshulThakur. @ssaishruthi98 , in addition to solution not being modified please pass non-primitive types as references to avoid copying objects around. Also, pass input arguments that you have no intention of changing as `const`.

With the current setup, I would guess that the limiter would not currently work? If it doesn't work, can you look into why your test would still pass? Ideally, we want the tests to catch this type of breaking modification.

Reply:
`solution` is now passed by reference. The tests did fail in this case, I just failed to catch that, my bad! Will ensure all limiter tests pass before committing new changes from next time.